File: server/game.py
import random
import unittest

import rules

DEBUG = False

TILES = rules.ALL_TILES * 4
PLAYER_TILES = 34
DISCARDS = 17

SEAT_WINDS = ('X1', 'X3')

def dummy_callback(player, msg_type, msg):
    import pprint
    pprint.pprint((player, msg_type, msg))

class Game(object):
    # Time limits, in seconds
    DISCARD_TIME_LIMIT = 30
    HAND_TIME_LIMIT = 3*60

    # Additional leeway to accomodate connection problems and UI updates
    EXTRA_TIME = 10

    def __init__(self,
                 nicks=['P1','P2'],
                 east=0,
                 callback=dummy_callback):
        self.callback = callback

        all_tiles = list(TILES)

        if not DEBUG:
            random.shuffle(all_tiles)

        self.nicks = nicks
        self.east = east

        # Tiles available for players
        n = PLAYER_TILES
        self.tiles = [all_tiles[:n], all_tiles[n:n*2]]

        self.dora_ind = all_tiles[n*2]
        self.uradora_ind = all_tiles[n*2+1]

        # Players' hands (None until they've chosen them)
        self.hand = [None, None]

        self.waits = [None, None]

        self.discards = [[], []]

        # Elapsed time in seconds (see beat() for details)
        self.t = 0
        self.deadlines = [None, None]

        self.finished = False

    @property
    def phase(self):
        # First phase - hand selection
        if not (self.hand[0] and self.hand[1]):
            return 1
        # Second phase - discards
        elif not self.finished:
            return 2
        # Game finished
        else:
            return 3

    @property
    def player_turn(self):
        if len(self.discards[0]) == len(self.discards[1]):
            return 0
        else:
            return 1

    def abort(self, culprit, description):
        '''Abort the game and provide explanation.'''

        for i in xrange(2):
            self.callback(i, 'abort',
                          {'culprit': culprit, 'description': description})
        self.finished = True
        self.deadlines = [None, None]

    def beat(self):
        '''Make time advance by 1 second in the game, handling possible timeouts.
        To be called by external code. Optional.'''
        self.t += 1
        for i in xrange(2):
            if self.deadlines[i] is not None and self.t >= self.deadlines[i]:
                self.abort(i, 'time limit exceeded')
                return

    def set_time_limit(self, player, time_limit):
        assert self.deadlines[player] is None
        self.deadlines[player] = self.t + time_limit + self.EXTRA_TIME

    def clear_time_limit(self, player):
        self.deadlines[player] = None

    def start(self):
        for i in xrange(2):
            self.callback(i, 'phase_one',
                          {'nicks': self.nicks,
                           'tiles': self.tiles[i],
                           'dora_ind': self.dora_ind,
                           'you': i,
                           'east': self.east})
            self.set_time_limit(i, self.HAND_TIME_LIMIT)

    def on_hand(self, player, hand):
        if self.phase != 1:
            self.abort(player, 'on_hand: wrong phase')
            return
        if len(hand) != 13:
            self.abort(player, 'on_hand: len != 13')
            return
        if self.hand[player] != None:
            self.abort(player, 'on_hand: hand already sent')
            return
        for tile in hand:
            # this fails if a player doesn't have a tile
            try:
                self.tiles[player].remove(tile)
            except ValueError:
                self.abort(player, 'on_hand: tile not found in choices')
                return

        self.clear_time_limit(player)

        self.hand[player] = hand
        self.waits[player] = list(rules.waits(hand))

        if self.hand[0] and self.hand[1]:
            # start the second phase
            for i in xrange(2):
                self.callback(i, 'phase_two', {})
            self.callback(self.player_turn, 'your_move', {})
            self.set_time_limit(self.player_turn, self.DISCARD_TIME_LIMIT)
        else:
            self.callback(player, 'wait_for_phase_two', {})

    def options(self, player, uradora=False):
        return {
            'fanpai_winds': [SEAT_WINDS[player^self.east]],
            'dora_ind': self.dora_ind,
            'uradora_ind': self.uradora_ind if uradora else None,
            'hotei': all(len(self.discards[i]) == DISCARDS for i in xrange(2)),

        'ippatsu': len(self.discards[1-player]) == 1,
        }

    def furiten(self, player):
        tiles = set(self.discards[player] + self.discards[1-player][:-1])
        return any(wait in tiles for wait in self.waits[player])

    def on_discard(self, player, tile):
        if self.phase != 2:
            self.abort(player, 'on_discard: wrong phase')
            return
        if self.player_turn != player:
            self.abort(player, 'on_discard: not your turn')
            return
        if tile not in self.tiles[player]:
            self.abort(player, 'on_discard: tile not found in choices')
            return

        self.clear_time_limit(player)
        self.tiles[player].remove(tile)
        self.discards[player].append(tile)

        for i in xrange(2):
            self.callback(i, 'discarded',
                          {'player': player,
                           'tile': tile})

        # ron
        if tile in self.waits[1-player] and not self.furiten(1-player):
            if self.check_ron(player, tile):
                return

        # draw
        if len(self.discards[0]) == len(self.discards[1]) == DISCARDS:
            self.finished = True
            for i in xrange(2):
                self.callback(i, 'draw', {})
        # normal turn
        else:
            self.callback(self.player_turn, 'your_move', {})
            self.set_time_limit(self.player_turn, self.DISCARD_TIME_LIMIT)

    def check_ron(self, player, tile):
        full_hand = sorted(self.hand[1-player] + [tile])
        hand = rules.best_hand(full_hand, tile, options=self.options(1-player))
        # Check mangan limit
        if hand.limit() == 0:
            return False

        # Compute again, with uradora
        hand = rules.best_hand(full_hand, tile,
                               options=self.options(1-player, uradora=True))
        self.finished = True
        for i in xrange(2):
            self.callback(i, 'ron', {
                'player': 1-player,
                'hand': full_hand,
                'yaku': hand.yaku,
                'yakuman': hand.yakuman,
                'dora': hand.dora(),
                'points': rules.BASE_POINTS[hand.limit()],
                'limit': hand.limit(),
                'uradora_ind': self.uradora_ind,
            })

        return True


class GameTestCase(unittest.TestCase):
    def setUp(self):
        from collections import deque

        global DEBUG
        DEBUG = True # suppress randomization

        self.messages = deque()

        self.g = Game(callback=self.callback)
        self.g.start()

    def assertMessage(self, player, msg_type, msg={}):
        our_m = self.messages.popleft()
        self.assertEqual(our_m, (player, msg_type, msg))

    def assertMessageBoth(self, msg_type, msg={}):
        self.assertMessage(0, msg_type, msg)
        self.assertMessage(1, msg_type, msg)

    def callback(self, player, msg_type, msg):
        self.messages.append((player, msg_type, msg))

    def test_init(self):
        # the game has just been created
        n = PLAYER_TILES
        self.assertMessage(0, 'phase_one',
                           {'nicks': ['P1', 'P2'],
                            'tiles': TILES[:n],
                            'dora_ind': 'M1',
                            'you': 0,
                            'east': 0})
        self.assertMessage(1, 'phase_one',
                           {'nicks': ['P1', 'P2'],
                            'tiles': TILES[n:n*2],
                            'dora_ind': 'M1',
                            'you': 1,
                            'east': 0})

    def start_game(self, s1, s2):
        self.test_init()
        self.g.on_hand(0, s1.split())
        self.assertMessage(0, 'wait_for_phase_two')
        self.g.on_hand(1, s2.split())
        self.assertMessageBoth('phase_two')

    def test_draw_scenario(self):
        self.start_game('M1 M2 M3 M4 M5 M6 M7 M8 M9 P1 P2 P3 P4',
                        'M1 M2 M3 M4 M5 M6 M7 M8 M9 P1 P2 P3 P4')

        for i in xrange(DISCARDS):
            for j in xrange(2):
                self.assertMessage(j, 'your_move')
                # Just discard the first choice
                t = self.g.tiles[j][0]
                self.discard(j, t)
        self.assertMessageBoth('draw')

    def discard(self, player, tile):
        self.g.on_discard(player, tile)
        self.assertMessageBoth('discarded', {'player': player,
                                             'tile': tile})

    def test_win(self):
        # P0: 13-sided kokushi
        # P1: riichi ippatsu dora - 3 fan, not enough for mangan
        self.start_game('M1 M9 P1 P9 S1 S9 X1 X2 X3 X4 X5 X6 X7',
                        'M1 M2 M3 M4 M5 M6 P7 P8 P9 S1 S2 S3 S4')
        self.assertMessage(0, 'your_move')

        # P1's winning tile - there should be no 'ron' message
        self.discard(0, 'S4')
        self.assertMessage(1, 'your_move')

        # Rising Sun!
        self.discard(1, 'P1')
        self.assertMessageBoth('ron', {
            'player': 0,
            'yaku': ['kokushi'],
            'yakuman': True,
            'hand': 'M1 M9 P1 P1 P9 S1 S9 X1 X2 X3 X4 X5 X6 X7'.split(),
            'points': 4000,
            'limit': 5,
            'dora': 0,
            'uradora_ind': 'M2'
        })


    def test_furiten(self):
        # P0: junk
        # P1: riichi tanyao sanshoku on S5, but only riichi tanyao on S8
        self.start_game('M2 M9 P1 P9 S1 S9 X1 X2 X3 X4 X5 X6 X7',
                        'M6 M7 M8 P6 P7 P8 S2 S3 S4 S5 S6 S7 S8')
        self.assertMessage(0, 'your_move')

        # P1's winning tile - there should be no 'ron' message
        # riichi ippatsu tanyao - no mangan
        self.discard(0, 'S8')
        self.assertMessage(1, 'your_move')

        self.discard(1, 'P1')
        self.assertMessage(0, 'your_move')

        # P1 would win now, but she's in furiten
        self.discard(0, 'S5')
        self.assertMessage(1, 'your_move')

    def test_short_hand(self):
        self.test_init()
        self.g.on_hand(1, 'M1 M2 M3'.split())
        self.assertMessageBoth('abort', {'culprit': 1, 'description': 'on_hand: len != 13'})
        self.assertTrue(self.g.finished)

    def test_tiles_outside_initial(self):
        self.test_init()
        self.g.on_hand(0, ['M1']*13)
        self.assertMessageBoth('abort', {'culprit': 0, 'description': 'on_hand: tile not found in choices'})
        self.assertTrue(self.g.finished)

    def test_hand_time_limit(self):
        self.test_init()
        self.g.on_hand(0, 'M2 M9 P1 P9 S1 S9 X1 X2 X3 X4 X5 X6 X7'.split())
        self.assertMessage(0, 'wait_for_phase_two')
        for i in range(self.g.HAND_TIME_LIMIT + self.g.EXTRA_TIME):
            self.g.beat()
        self.assertMessageBoth('abort', {'culprit': 1, 'description': 'time limit exceeded'})

    def test_discard_time_limit(self):
        self.start_game('M1 M2 M3 M4 M5 M6 M7 M8 M9 P1 P2 P3 P4',
                        'M1 M2 M3 M4 M5 M6 M7 M8 M9 P1 P2 P3 P4')
        self.assertMessage(0, 'your_move')
        for i in range(self.g.DISCARD_TIME_LIMIT + self.g.EXTRA_TIME):
            self.g.beat()
        self.assertMessageBoth('abort', {'culprit': 0, 'description': 'time limit exceeded'})


if __name__ == '__main__':
    unittest.main()

File: server-rs/minefield-game/src/game.rs
use rand::seq::SliceRandom;

use minefield_core::score::Score;
use minefield_core::search::{find_all_waits, search};
use minefield_core::tiles::Tile;
use minefield_core::yaku;
use yaku::Yaku;

use crate::protocol::{MoveType, Msg};

const PLAYER_TILES: usize = 34;
const DISCARDS: usize = 17;
const DISCARD_TIME_LIMIT: usize = 15;
const HAND_TIME_LIMIT: usize = 3 * 60;
const EXTRA_TIME: usize = 10;

pub struct Game {
    east: usize,
    players: [Player; 2],
    dora_ind: Tile,
    uradora_ind: Tile,
    finished: bool,
    time: usize,
    messages: Vec<(usize, Msg)>,
}

pub enum GameError {
    Abort { culprit: usize, description: usize },
}

impl Game {
    pub fn new(rng: &mut impl rand::Rng) -> Self {
        let mut all_tiles = all_tiles();
        all_tiles.shuffle(rng);

        let east = rng.gen_range(0, 2);

        Self::fixed(&all_tiles, east)
    }

    pub fn fixed(all_tiles: &[Tile], east: usize) -> Self {
        let players = [
            Player::new(&all_tiles[0..PLAYER_TILES], east == 0),
            Player::new(&all_tiles[PLAYER_TILES..PLAYER_TILES * 2], east == 1),
        ];

        let dora_ind = all_tiles[PLAYER_TILES * 2];
        let uradora_ind = all_tiles[PLAYER_TILES * 2 + 1];

        Game {
            east,
            players,
            dora_ind,
            uradora_ind,
            finished: false,
            time: 0,
            messages: vec![],
        }
    }

    pub fn on_start(&mut self) {
        for i in 0..2 {
            self.send(
                i,
                Msg::PhaseOne {
                    tiles: self.players[i].tiles.clone(),
                    dora_ind: self.dora_ind,
                    you: i,
                    east: self.east,
                },
            );
            self.start_move(i, MoveType::Hand);
        }
    }

    pub fn on_message(&mut self, i: usize, msg: Msg) {
        assert!(!self.finished);
        match msg {
            Msg::Hand { hand } => {
                self.on_hand(i, &hand);
            }
            Msg::Discard { tile } => {
                self.on_discard(i, tile);
            }
            _ => self.abort(i, "unexpected message"),
        }
    }

    pub fn messages(&mut self) -> Vec<(usize, Msg)> {
        self.messages.split_off(0)
    }

    pub fn messages_at_most(&mut self, n: usize) -> Vec<(usize, Msg)> {
        let suffix = self
            .messages
            .split_off(std::cmp::min(n, self.messages.len()));
        std::mem::replace(&mut self.messages, suffix)
    }

    pub fn beat(&mut self) {
        assert!(!self.finished);

        self.time += 1;
        for i in 0..2 {
            if let Some(deadline) = self.players[i].deadline {
                if self.time == deadline {
                    return self.abort(i, "time limit exceeded");
                }
            }
        }
    }

    fn send(&mut self, i: usize, msg: Msg) {
        self.messages.push((i, msg));
    }

    fn send_both(&mut self, msg: Msg) {
        self.messages.push((0, msg.clone()));
        self.messages.push((1, msg));
    }

    fn start_move(&mut self, i: usize, move_type: MoveType) {
        assert!(self.players[i].deadline.is_none());
        let time_limit = match move_type {
            MoveType::Hand => HAND_TIME_LIMIT,
            MoveType::Discard => DISCARD_TIME_LIMIT,
        };
        self.players[i].deadline = Some(self.time + time_limit + EXTRA_TIME);
        self.send(
            i,
            Msg::StartMove {
                move_type,
                time_limit,
            },
        )
    }

    fn end_move(&mut self, i: usize) {
        assert!(self.players[i].deadline.is_some());
        self.players[i].deadline = None;
        self.send(i, Msg::EndMove);
    }

    fn on_hand(&mut self, i: usize, hand: &[Tile]) {
        let player = &mut self.players[i];
        if let Err(description) = player.set_hand(hand) {
            return self.abort(i, description);
        }
        self.end_move(i);

        // For replay
        self.send(
            i,
            Msg::Hand {
                hand: hand.to_vec(),
            },
        );

        if self.is_phase2() {
            self.send_both(Msg::PhaseTwo);

            self.start_move(self.east, MoveType::Discard);
        } else {
            self.send(i, Msg::WaitForPhaseTwo);
        }
    }

    fn is_phase2(&self) -> bool {
        !self.players[0].hand.is_empty() && !self.players[1].hand.is_empty()
    }

    fn abort(&mut self, culprit: usize, description: &str) {
        self.finished = true;

        self.send_both(Msg::Abort {
            culprit,
            description: description.to_owned(),
        });
    }

    pub fn on_discard(&mut self, i: usize, tile: Tile) {
        if !self.is_phase2() {
            return self.abort(i, "discard too soon");
        }

        if let Err(description) = self.players[i].discard(tile) {
            return self.abort(i, description);
        }
        self.end_move(i);
        for j in 0..2 {
            self.send(j, Msg::Discarded { player: i, tile });
        }

        // ron
        if let Some(msg) =
            self.players[1 - i].check_ron(1 - i, tile, self.dora_ind, self.uradora_ind)
        {
            self.send_both(msg);
            self.finished = true;
            return;
        }

        // draw
        if self.players[0].finished() && self.players[1].finished() {
            self.send_both(Msg::Draw);
            self.finished = true;
            return;
        }

        // normal turn
        self.start_move(1 - i, MoveType::Discard);
    }
}

fn all_tiles() -> Vec<Tile> {
    let mut result = vec![];
    for _ in 0..4 {
        result.append(&mut Tile::all().collect());
    }
    result
}

struct Player {
    tiles: Vec<Tile>,
    is_east: bool,
    deadline: Option<usize>,
    hand: Vec<Tile>,
    discards: Vec<Tile>,
    furiten: bool,
    waits: Vec<Tile>,
}

impl Player {
    fn new(tiles: &[Tile], is_east: bool) -> Self {
        Player {
            tiles: tiles.to_vec(),
            is_east,
            deadline: None,
            hand: vec![],
            discards: vec![],
            furiten: false,
            waits: vec![],
        }
    }

    fn set_hand(&mut self, hand: &[Tile]) -> Result<(), &'static str> {
        if !self.deadline.is_some() || !self.hand.is_empty() {
            return Err("not expecting a hand");
        }
        if hand.len() != 13 {
            return Err("len != 13");
        }
        for tile in hand.iter() {
            let idx = self.find_choice(*tile)?;
            self.tiles.remove(idx);
        }
        let waits = find_all_waits(hand);

        self.hand = hand.to_vec();
        self.waits = waits;

        Ok(())
    }

    fn find_choice(&self, tile: Tile) -> Result<usize, &'static str> {
        self.tiles
            .iter()
            .position(|t| *t == tile)
            .ok_or("tile not found in choices")
    }

    fn discard(&mut self, tile: Tile) -> Result<(), &'static str> {
        if !self.deadline.is_some() || self.hand.is_empty() {
            return Err("not expecting a discard");
        }
        let idx = self.find_choice(tile)?;
        self.tiles.remove(idx);
        assert!(self.discards.len() < DISCARDS);
        self.discards.push(tile);
        if self.waits.contains(&tile) {
            self.furiten = true;
        }
        Ok(())
    }

    fn finished(&self) -> bool {
        self.discards.len() == DISCARDS
    }

    fn check_ron(
        &mut self,
        i: usize,
        tile: Tile,
        dora_ind: Tile,
        uradora_ind: Tile,
    ) -> Option<Msg> {
        if !self.waits.contains(&tile) {
            return None;
        }
        if self.furiten {
            return None;
        }

        let player_wind = if self.is_east { Tile::X1 } else { Tile::X3 };
        let dora = dora_ind.next_wrap();
        let uradora = uradora_ind.next_wrap();
        let turn = if self.is_east {
            self.discards.len() - 1
        } else {
            self.discards.len()
        };

        let mut full_hand = self.hand.clone();
        full_hand.push(tile);
        let hands = search(&full_hand, tile);
        let scored_hands = hands.iter().filter_map(|hand| {
            let mut special = vec![Yaku::Riichi];
            if turn == 0 {
                special.push(Yaku::Ippatsu);
            }
            if turn == DISCARDS - 1 {
                special.push(Yaku::Hotei);
            }
            let mut score = Score::from_hand(hand, player_wind, &special);

            // Check if mangan (with dora)
            score.add_dora(dora);
            if score.limit() == 0 {
                return None;
            }

            // Return score (with dora and uradora)
            score.add_dora(uradora);
            Some(score)
        });

        if let Some(score) = scored_hands.max_by_key(|score| score.limit()) {
            Some(Msg::Ron {
                player: i,
                hand: self.hand.clone(),
                tile,
                yaku: score.yaku.clone(),
                dora: score.dora_count,
                uradora_ind,
                limit: score.limit(),
                points: score.points(),
            })
        } else {
            self.furiten = true;
            None
        }
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use Tile::*;

    fn game() -> Game {
        Game::fixed(&all_tiles(), 0)
    }

    fn assert_aborted(game: &mut Game, culprit: usize, description: &str) {
        let abort = Msg::Abort {
            culprit,
            description: description.to_owned(),
        };
        assert_eq!(game.messages(), vec![(0, abort.clone()), (1, abort)]);
    }

    #[test]
    fn test_init() {
        let mut game = game();
        let all_tiles = all_tiles();
        game.on_start();
        assert_eq!(
            game.messages(),
            vec![
                (
                    0,
                    Msg::PhaseOne {
                        tiles: all_tiles[0..PLAYER_TILES].to_vec(),
                        dora_ind: M1,
                        you: 0,
                        east: 0
                    }
                ),
                (
                    0,
                    Msg::StartMove {
                        move_type: MoveType::Hand,
                        time_limit: HAND_TIME_LIMIT
                    }
                ),
                (
                    1,
                    Msg::PhaseOne {
                        tiles: all_tiles[PLAYER_TILES..2 * PLAYER_TILES].to_vec(),
                        dora_ind: M1,
                        you: 1,
                        east: 0
                    }
                ),
                (
                    1,
                    Msg::StartMove {
                        move_type: MoveType::Hand,
                        time_limit: HAND_TIME_LIMIT
                    }
                )
            ]
        );
    }

    fn start_game(hand_0: &[Tile], hand_1: &[Tile]) -> Game {
        let mut game = game();
        game.on_start();
        game.messages();
        game.on_message(
            0,
            Msg::Hand {
                hand: hand_0.to_vec(),
            },
        );
        assert_eq!(
            game.messages(),
            vec![
                (0, Msg::EndMove),
                (
                    0,
                    Msg::Hand {
                        hand: hand_0.to_vec()
                    }
                ),
                (0, Msg::WaitForPhaseTwo)
            ]
        );
        game.on_message(
            1,
            Msg::Hand {
                hand: hand_1.to_vec(),
            },
        );
        assert_eq!(
            game.messages_at_most(4),
            vec![
                (1, Msg::EndMove),
                (
                    1,
                    Msg::Hand {
                        hand: hand_1.to_vec()
                    }
                ),
                (0, Msg::PhaseTwo),
                (1, Msg::PhaseTwo),
            ]
        );
        game
    }

    fn discard(game: &mut Game, player: usize, tile: Tile) {
        assert_eq!(
            game.messages(),
            vec![(
                player,
                Msg::StartMove {
                    move_type: MoveType::Discard,
                    time_limit: DISCARD_TIME_LIMIT
                }
            )]
        );
        game.on_message(player, Msg::Discard { tile });
        assert_eq!(
            game.messages_at_most(3),
            vec![
                (player, Msg::EndMove),
                (0, Msg::Discarded { player, tile }),
                (1, Msg::Discarded { player, tile }),
            ]
        );
    }

    #[test]
    fn test_draw() {
        let mut game = start_game(
            &[M1, M2, M3, M4, M5, M6, M7, M8, M9, P1, P2, P3, P4],
            &[M1, M2, M3, M4, M5, M6, M7, M8, M9, P1, P2, P3, P4],
        );
        for _ in 0..DISCARDS {
            for player in 0..2 {
                let tile = game.players[player].tiles[0];
                discard(&mut game, player, tile);
            }
        }
        assert_eq!(game.messages(), vec![(0, Msg::Draw), (1, Msg::Draw)]);
        assert_eq!(game.finished, true);
    }

    #[test]
    fn test_win() {
        // P0: 13-sided kokushi
        // P1: riichi ippatsu dora - 3 fan, not enough for mangan
        let mut game = start_game(
            &[M1, M9, P1, P9, S1, S9, X1, X2, X3, X4, X5, X6, X7],
            &[M1, M2, M3, M4, M5, M6, P7, P8, P9, S1, S2, S3, S4],
        );

        // P1's winning tile - no ron, furiten
        discard(&mut game, 0, S4);
        assert_eq!(game.finished, false);
        assert_eq!(game.players[1].furiten, true);

        // Rising Sun!
        discard(&mut game, 1, P1);
        let ron = Msg::Ron {
            player: 0,
            yaku: vec![Yaku::Kokushi],
            hand: vec![M1, M9, P1, P9, S1, S9, X1, X2, X3, X4, X5, X6, X7],
            points: 32000,
            limit: 5,
            dora: 0,
            uradora_ind: M2,
            tile: P1,
        };
        assert_eq!(game.messages(), vec![(0, ron.clone()), (1, ron)]);
        assert_eq!(game.finished, true);
    }

    #[test]
    fn test_furiten() {
        // P0: junk
        // P1: riichi tanyao sanshoku on S5, but only riichi tanyao on S8
        let mut game = start_game(
            &[M2, M9, P1, P9, S1, S9, X1, X2, X3, X4, X5, X6, X7],
            &[M6, M7, M8, P6, P7, P8, S2, S3, S4, S5, S6, S7, S8],
        );

        // P1's winning tile - no ron, furiten
        discard(&mut game, 0, S8);
        assert_eq!(game.finished, false);
        assert_eq!(game.players[1].furiten, true);

        discard(&mut game, 1, P1);

        // P1 would win now, but she's in furiten
        discard(&mut game, 0, S5);
        assert_eq!(game.finished, false);
    }

    #[test]
    fn test_short_hand() {
        let mut game = game();
        game.on_start();
        game.messages();
        game.on_message(
            1,
            Msg::Hand {
                hand: vec![M1, M2, M3],
            },
        );
        assert_aborted(&mut game, 1, "len != 13");
    }

    #[test]
    fn test_tiles_outside_initial() {
        let mut game = game();
        game.on_start();
        game.messages();
        game.on_message(
            1,
            Msg::Hand {
                hand: vec![M1, M1, M1, M1, M1, M1, M1, M1, M1, M1, M1, M1, M1],
            },
        );
        assert_aborted(&mut game, 1, "tile not found in choices");
    }

    #[test]
    fn test_hand_time_limit() {
        let mut game = game();
        game.on_start();
        game.messages();
        game.on_message(
            0,
            Msg::Hand {
                hand: vec![M2, M9, P1, P9, S1, S9, X1, X2, X3, X4, X5, X6, X7],
            },
        );
        game.messages();
        for _ in 0..HAND_TIME_LIMIT + EXTRA_TIME {
            game.beat();
        }
        assert_aborted(&mut game, 1, "time limit exceeded");
    }

    #[test]
    fn test_discard_time_limit() {
        let mut game = start_game(
            &[M1, M2, M3, M4, M5, M6, M7, M8, M9, P1, P2, P3, P4],
            &[M1, M2, M3, M4, M5, M6, M7, M8, M9, P1, P2, P3, P4],
        );
        game.messages();
        for _ in 0..DISCARD_TIME_LIMIT + EXTRA_TIME {
            game.beat();
        }
        assert_aborted(&mut game, 0, "time limit exceeded");
    }
}
